File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        t, d = 0 , len(matrix) - 1
        while t <= d:
            row = (t + d) // 2
            if matrix[row][0] <= target <= matrix[row][-1]:
                break
            if target < matrix[row][0]:
                d = row - 1
            else:
                t = row + 1
        
        l, r = 0, len(matrix[0]) - 1
        while l <= r:
            m = (l + r) // 2
            if matrix[row][m] == target:
                return True
            if target < matrix[row][m]:
                r = m - 1
            else:
                l = m + 1
        return False
            
        

File: Data Structures & Algorithms/set-zeroes-in-matrix/submission-1.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        set<int> zero_rows, zero_cols;
        int ROWS = matrix.size(), COLS = matrix[0].size();

        for (int r = 0; r < ROWS; r++){
            for (int c = 0; c < COLS; c++){
                if (matrix[r][c] == 0){
                    zero_rows.insert(r);
                    zero_cols.insert(c);
                }
            }
        }
        for (int r = 0; r < ROWS; r++){
            for (int c = 0; c < COLS; c++){
                if (zero_rows.count(r) || 
                zero_cols.count(c)) matrix[r][c] = 0;
            }
        }
        
    }
};
